Tabular: Per model HPO control

Innixma opened the pull request:
*Issue #, if available:*

*Description of changes:*

- Enables per model HPO control
- Now each model can specify its own scheduler (or lack of one) as an `ag_args` value.
- Enabled usage of 'custom' models when HPO is enabled. (custom models have HPO disabled so they just train normally)

Example:

```
hyperparameters = {
        'GBM': {'ag_args': {'hyperparameter_tune_kwargs': None}},
        'CAT': {'ag_args': {'hyperparameter_tune_kwargs': 'bayesopt'}},
        'NN': {},
    }
```

In this example:

1. GBM model will not perform any HPO because of `{'hyperparameter_tune_kwargs': None}`
2. CAT model will perform HPO with bayesopt
3. NN model will perform HPO based on the predictor.fit `hyperparameter_tune_kwargs` value (or not do HPO if not specified).
4. predictor.fit `hyperparameter_tune_kwargs` value acts as the value in predictor.fit `ag_args['hyperparameter_tune_kwargs']`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

szha commented:
Job PR-883-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-883/1/index.html

jwmueller commented on tabular/src/autogluon/tabular/trainer/model_presets/presets.py:
```diff
@@ -112,121 +110,168 @@
-        model_names_set.update(invalid_model_names)
+
+    if not silent:
+        logger.log(20, 'Model configs to train:')
```
may be more clear: "to train" -> "that will be trained". 
Also should this this info be dumped by default, or should it have slightly lower logging-level priority so it's only output under higher verbosity?

Innixma replied:
Updated. This is controlled by silent argument, which is toggled off if verbosity > 2 in Trainer.

jwmueller commented on tabular/src/autogluon/tabular/trainer/model_presets/presets.py:
```diff
@@ -112,121 +110,168 @@
+        if 'hyperparameter_tune_kwargs' in model[AG_ARGS]:
+            model_args_fit[model_init.name] = {'hyperparameter_tune_kwargs': model[AG_ARGS]['hyperparameter_tune_kwargs']}
+        if not silent:
+            logger.log(20, f'\t{model_init.name}: \t{model}')
```
I'd recommend additional more detailed logging statement (at lower logging level) for higher-verbosity users. Since the initial statement talk about the model-configs that will be trained, it could be good at higher verbosity to actually print these configs (ie. their non-default ag_args, ag_args_fit values, non-default hyperparameter-values, etc) before training commences

Innixma replied:
This actually is the detailed log, the variable names were just a bit confusing. Updated. This is only printed if verbosity >= 3.

Example:

```
Model configs that will be trained:
	RandomForestGini: 	{'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass'], 'model_type': <class 'autogluon.tabular.models.rf.rf_model.RFModel'>, 'priority': 100}}
	RandomForestEntr: 	{'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass'], 'model_type': <class 'autogluon.tabular.models.rf.rf_model.RFModel'>, 'priority': 100}}
	ExtraTreesGini: 	{'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass'], 'model_type': <class 'autogluon.tabular.models.xt.xt_model.XTModel'>, 'priority': 90}}
	ExtraTreesEntr: 	{'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass'], 'model_type': <class 'autogluon.tabular.models.xt.xt_model.XTModel'>, 'priority': 90}}
	KNeighborsUnif: 	{'weights': 'uniform', 'ag_args': {'name_suffix': 'Unif', 'model_type': <class 'autogluon.tabular.models.knn.knn_model.KNNModel'>, 'priority': 80}}
	KNeighborsDist: 	{'weights': 'distance', 'ag_args': {'name_suffix': 'Dist', 'model_type': <class 'autogluon.tabular.models.knn.knn_model.KNNModel'>, 'priority': 80}}
	LightGBM: 	{'ag_args': {'model_type': <class 'autogluon.tabular.models.lgb.lgb_model.LGBModel'>, 'priority': 70}}
	LightGBMXT: 	{'extra_trees': True, 'ag_args': {'name_suffix': 'XT', 'model_type': <class 'autogluon.tabular.models.lgb.lgb_model.LGBModel'>, 'priority': 70}}
	CatBoost: 	{'ag_args': {'model_type': <class 'autogluon.tabular.models.catboost.catboost_model.CatBoostModel'>, 'priority': 60}}
	XGBoost: 	{'ag_args': {'model_type': <class 'autogluon.tabular.models.xgboost.xgboost_model.XGBoostModel'>, 'priority': 55}}
	NeuralNetMXNet: 	{'ag_args': {'model_type': <class 'autogluon.tabular.models.tabular_nn.tabular_nn_model.TabularNeuralNetModel'>, 'priority': 50}}
	NeuralNetFastAI: 	{'ag_args': {'model_type': <class 'autogluon.tabular.models.fastainn.tabular_nn_fastai.NNFastAiTabularModel'>, 'priority': 45}}
	LightGBMCustom: 	{'num_boost_round': 10000, 'num_threads': -1, 'objective': 'binary', 'verbose': -1, 'boosting_type': 'gbdt', 'learning_rate': 0.03, 'num_leaves': 128, 'feature_fraction': 0.9, 'min_data_in_leaf': 5, 'two_round': True, 'seed_value': 0, 'ag_args': {'model_type': <class 'autogluon.tabular.models.lgb.lgb_model.LGBModel'>, 'name_suffix': 'Custom', 'disable_in_hpo': True, 'hyperparameter_tune_kwargs': None, 'priority': 0}}
```


Example with HPO:

```
Model configs that will be trained:
	LightGBM: 	{'ag_args': {'hyperparameter_tune_kwargs': None, 'model_type': <class 'autogluon.tabular.models.lgb.lgb_model.LGBModel'>, 'priority': 70}}
	LightGBM_2: 	{'ag_args': {'hyperparameter_tune_kwargs': (<class 'autogluon.core.scheduler.fifo.FIFOScheduler'>, {'resource': {'num_cpus': 'auto', 'num_gpus': 'auto'}, 'searcher': 'random', 'search_options': {}, 'checkpoint': None, 'resume': False, 'num_trials': None, 'reward_attr': 'validation_performance', 'time_attr': 'epoch', 'visualizer': 'none', 'dist_ip_addrs': []}), 'model_type': <class 'autogluon.tabular.models.lgb.lgb_model.LGBModel'>, 'priority': 70}}
	CatBoost: 	{'ag_args': {'hyperparameter_tune_kwargs': 'bayesopt', 'model_type': <class 'autogluon.tabular.models.catboost.catboost_model.CatBoostModel'>, 'priority': 60}}

```

jwmueller commented:
Are we checking somewhere that all of the `ag_args*` arguments only contain valid keys? Seems easy for user to specify wrong keys in these on accident

jwmueller commented:
Probably need to include example of using `hyperparameter_tune_kwargs` this in documentation for `ag_args`, otherwise this functionality will be hidden.  
Also user may be confused about:
1) why this is not `ag_args_fit` argument whereas eg. time-limit is (perhaps need to clarify `ag_args` description).
2) what are the options for `hyperparameter_tune_kwargs` key in `ag_args`

jwmueller commented:
`_init_scheduler_tabular` should have at least one line docstring stating in plain English what the HPO strategy is now (similar to how before we had the comment: all models use same scheduler / HPO-strategy). Otherwise it's quite complicated for users/devs to figure out what is happening even just at a high-level.

jwmueller reviewed (changes requested):
Probably needs more docs / clarifications to enable understanding of the new HPO-functionality / options available (otherwise may add to user-confusion)

gradientsky commented:
How is scheduler options will be passed if different schedulers are used?

Innixma commented:
> Are we checking somewhere that all of the `ag_args*` arguments only contain valid keys? Seems easy for user to specify wrong keys in these on accident

We were not, I have changed this and now it is checked and a warning is printed if an unknown ag_args is passed. ag_args_fit and ag_args_ensemble don't fit this requirement nicely since each model could theoretically use their own values.

Innixma commented:
> `_init_scheduler_tabular` should have at least one line docstring stating in plain English what the HPO strategy is now (similar to how before we had the comment: all models use same scheduler / HPO-strategy). Otherwise it's quite complicated for users/devs to figure out what is happening even just at a high-level.

Changed this code around to be cleaner and delay construction of scheduler until later on.

Innixma commented:
> Probably need to include example of using `hyperparameter_tune_kwargs` this in documentation for `ag_args`, otherwise this functionality will be hidden.

Added

> Also user may be confused about:
> 
> 1. why this is not `ag_args_fit` argument whereas eg. time-limit is (perhaps need to clarify `ag_args` description).

Improved ag_args description. Reason is that ag_args_fit applies to a singular model fit, whereas tune_kwargs controls things outside of model fit. time_limit in ag_args_fit is specific to a single model fit, so it is not outside of fit scope.

> 2. what are the options for `hyperparameter_tune_kwargs` key in `ag_args`. Related to: Why do we still need to accept `hyperparameter_tune_kwargs` as a kwarg of tabular.fit()?  Shouldn't the user only pass it now via `ag_args['hyperparameter_tune_kwargs']` instead? Offering too many ways for the user to pass this arg may also confuse them.

This is an open question, ag_args['hyperparameter_tune_kwargs'] and hyperparameter_tune_kwargs are the same. I have added an exception if both are specified, and added documentation to ag_args for hyperparameter_tune_kwargs.

I believe it makes sense to keep both options since other tasks will have hyperparameter_tune_kwargs as an argument.

Innixma commented:
> Can you add an example where different schedulers are used with different scheduler parameters (useful if different schedulers are used with different sets of parameters)?

```

import pandas as pd
import sklearn.datasets

from autogluon.tabular import TabularPredictor
from autogluon.tabular.utils import generate_train_test_split

LABEL = 'target'


if __name__ == '__main__':
    cancer_data = sklearn.datasets.load_breast_cancer(return_X_y=False)
    X = pd.DataFrame(cancer_data['data'], columns=cancer_data['feature_names'])
    X[LABEL] = cancer_data[LABEL]

    X_train, X_test, y_train, y_test = generate_train_test_split(X=X.drop(LABEL, axis=1), y=X[LABEL], problem_type='binary', test_size=0.2)

    X_train[LABEL] = y_train

    hyperparameters = {
        'GBM': [
            {'ag_args': {'hyperparameter_tune_kwargs': None}},
            {},
            {'ag_args': {'hyperparameter_tune_kwargs': {'searcher': 'random', 'num_trials': 3}}},
            {'ag_args': {'hyperparameter_tune_kwargs': {'searcher': 'random', 'resource': {'num_cpus': 1}, 'time_out': 20}}},
        ],
        'CAT': {'ag_args': {'hyperparameter_tune_kwargs': 'bayesopt'}},
        'custom': ['GBM'],
    }

    predictor = TabularPredictor(
        label=LABEL,
        verbosity=3,
    ).fit(
        train_data=X_train,
        time_limit=120,
        hyperparameters=hyperparameters,
        hyperparameter_tune_kwargs='auto',
    )

    X_test[LABEL] = y_test
    predictor.leaderboard(X_test)

    predictor.fit_extra(hyperparameters={'XGB': {}}, hyperparameter_tune_kwargs='auto', time_limit=20)

    predictor.leaderboard(X_test)

```

szha commented:
Job PR-883-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-883/5/index.html

jwmueller commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -289,17 +286,18 @@ def fit(self,
+                            disable_in_hpo: (bool) If True, the model will only be trained if `hyperparameter_tune_kwargs=None`.
                             valid_stacker: (bool) If False, the model will not be trained as a level 1 or higher stacker model.
                             valid_base: (bool) If False, the model will not be trained as a level 0 (base) model.
+                            hyperparameter_tune_kwargs: (dict) Refer to :meth:`TabularPredictor.fit` hyperparameter_tune_kwargs argument. If specified here, will override global HPO settings for this model.
```
"for this model" -> "for just this model (so you can tune different models via different HPO search strategies)"

jwmueller reviewed (approved):
looks good, just suggested one more minor clarification.

"for this model" -> "for just this model (so you can tune different models via different HPO search strategies)"
